network/stream/v2/syncing_test.go

```diff
@@ -0,0 +1,811 @@
+				// and divide it by half to perform a check earlier.
+				sleep = time.Duration(float64(time.Since(start)) * float64(want-count) / float64(count) / 2)
+				log.Debug("expecting all chunks", "in", sleep*2, "want", want, "have", count)
+			}
+		}
+		switch {
+		case sleep > time.Minute:
+			// next check and speed calculation in some shorter time
+			sleep = 500 * time.Millisecond
+		case sleep > 5*time.Second:
+			// upper limit for the check, do not check too slow
+			sleep = 5 * time.Second
+		case sleep < 50*time.Millisecond:
+			// lower limit for the check, do not check too frequently
+			sleep = 50 * time.Millisecond
+			if staled > 0 {
+				// slow down if chunks are stuck near the want value
+				sleep *= 10
+			}
+		}
+		time.Sleep(sleep)
+	}
+
+	if count != want {
+		return fmt.Errorf("got synced chunks %d, want %d", count, want)
+	}
+	return nil
+}
+
+func getChunkCount(store chunk.Store) (c uint64, err error) {
+	for po := 0; po <= chunk.MaxPO; po++ {
+		last, err := store.LastPullSubscriptionBinID(uint8(po))
+		if err != nil {
+			return 0, err
+		}
+		c += last
+	}
+	return c, nil
+}
+
+func getChunks(store chunk.Store) (chunks map[string]struct{}, err error) {
+	chunks = make(map[string]struct{})
+	for po := uint8(0); po <= chunk.MaxPO; po++ {
+		last, err := store.LastPullSubscriptionBinID(po)
+		if err != nil {
+			return nil, err
+		}
+		if last == 0 {
+			continue
+		}
+		ch, _ := store.SubscribePull(context.Background(), po, 0, last)
+		for c := range ch {
+			addr := c.Address.Hex()
+			if _, ok := chunks[addr]; ok {
+				return nil, fmt.Errorf("duplicate chunk %s", addr)
+			}
+			chunks[addr] = struct{}{}
+		}
+	}
+	return chunks, nil
+}
+
+/*
+ go test -v -bench . -run BenchmarkHistoricalStream -loglevel 0 -benchtime 10x
+BenchmarkHistoricalStream_1000-4    	      10	 119487009 ns/op
+BenchmarkHistoricalStream_2000-4    	      10	 236469752 ns/op
+BenchmarkHistoricalStream_3000-4    	      10	 371934729 ns/op
+BenchmarkHistoricalStream_5000-4    	      10	 638317966 ns/op
+BenchmarkHistoricalStream_10000-4   	      10	1359858063 ns/op
+BenchmarkHistoricalStream_15000-4   	      10	2485790336 ns/op
+BenchmarkHistoricalStream_20000-4   	      10	3382260295 ns/op
+*/
+func BenchmarkHistoricalStream_1000(b *testing.B)  { benchmarkHistoricalStream(b, 1000) }
+func BenchmarkHistoricalStream_2000(b *testing.B)  { benchmarkHistoricalStream(b, 2000) }
+func BenchmarkHistoricalStream_3000(b *testing.B)  { benchmarkHistoricalStream(b, 3000) }
+func BenchmarkHistoricalStream_5000(b *testing.B)  { benchmarkHistoricalStream(b, 5000) }
+func BenchmarkHistoricalStream_10000(b *testing.B) { benchmarkHistoricalStream(b, 10000) }
+func BenchmarkHistoricalStream_15000(b *testing.B) { benchmarkHistoricalStream(b, 15000) }
+func BenchmarkHistoricalStream_20000(b *testing.B) { benchmarkHistoricalStream(b, 20000) }
+
+func benchmarkHistoricalStream(b *testing.B, chunks uint64) {
```

Comment:
@zelig @nonsense Could you check this benchmark also? I think that the measuring method is not correct. The simulation is constructed outside of the benchmark loop and in the benchmark loop chunks are uploaded on the same node of the same simulation, making the state on every benchmark iteration different. Every iteration measures different state and with larger number of chunks in the localstore, it becomes slower. This makes this benchmark result dependent on the number of iterations performed, where only the precisions of results should be influenced by the number of iterations. I think that this benchmark should be rewritten in a way that every iteration measures time on the independent state of other iterations, or to remove the benchmark as that may make this benchmark time consuming.

Reply:
@janos it does look like the benchmark is indeed not correct for the reasons you mentioned.

Reply:
agree

Reply:
I have changed the benchmark to address this problem, and it is even usable.